Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp

```diff
@@ -0,0 +1,199 @@
+        if (m_state == State::Canceling || m_state == State::Completed)
+            return;
+
+        m_dataURLResult = WTFMove(decodeResult);
+        didDecodeDataURL();
+    });
+}
+
+void NetworkDataTaskDataURL::cancel()
+{
+    if (m_state == State::Canceling || m_state == State::Completed)
+        return;
+
+    m_state = State::Canceling;
+}
+
+void NetworkDataTaskDataURL::invalidateAndCancel()
+{
+    cancel();
+    m_state = State::Completed;
+}
+
+NetworkDataTask::State NetworkDataTaskDataURL::state() const
+{
+    return m_state;
+}
+
+void NetworkDataTaskDataURL::setPendingDownloadLocation(const String& filename, SandboxExtension::Handle&& sandboxExtensionHandle, bool allowOverwrite)
+{
+    NetworkDataTask::setPendingDownloadLocation(filename, WTFMove(sandboxExtensionHandle), allowOverwrite);
+    m_allowOverwriteDownload = allowOverwrite;
+}
+
+String NetworkDataTaskDataURL::suggestedFilename() const
+{
+    if (!m_suggestedFilename.isEmpty())
+        return m_suggestedFilename;
+
+    String suggestedFilename = m_response.suggestedFilename();
+    if (!suggestedFilename.isEmpty())
+        return suggestedFilename;
+
+    return PAL::decodeURLEscapeSequences(m_response.url().lastPathComponent());
+}
+
+void NetworkDataTaskDataURL::didDecodeDataURL()
+{
+    ASSERT(m_state == State::Running);
+    ASSERT(m_decodingState == DecodingState::Finished);
+    if (!m_dataURLResult) {
+        if (m_client)
+            m_client->didCompleteWithError(internalError(firstRequest().url()));
+        return;
+    }
+
+    m_response = ResourceResponse::dataURLResponse(firstRequest().url(), m_dataURLResult.value());
+
+    didReceiveResponse(ResourceResponse(m_response), NegotiatedLegacyTLS::No, PrivateRelayed::No, [this, protectedThis = Ref { *this }](PolicyAction policyAction) {
+        if (m_state == State::Canceling || m_state == State::Completed)
+            return;
+
+        switch (policyAction) {
+        case PolicyAction::Use:
+            // Should not be reached for data URLs as they are normally handled in the web process.
+            invalidateAndCancel();
+            ASSERT_NOT_REACHED();
+            break;
+        case PolicyAction::Ignore:
+            invalidateAndCancel();
+            break;
+        case PolicyAction::Download:
+            downloadDecodedData();
+            break;
+        case PolicyAction::StopAllLoads:
+            ASSERT_NOT_REACHED();
+            break;
+        }
+    });
+}
+
+void NetworkDataTaskDataURL::downloadDecodedData()
+{
+    FileSystem::PlatformFileHandle downloadDestinationFile = FileSystem::openFile(m_pendingDownloadLocation, FileSystem::FileOpenMode::Truncate, FileSystem::FileAccessPermission::All, !m_allowOverwriteDownload);
+    if (!FileSystem::isHandleValid(downloadDestinationFile)) {
+#if USE(CURL)
+        ResourceError error(CURLE_WRITE_ERROR, m_response.url());
```

Comment:
CURLE_WRITE_ERROR does not represent the correct error. If there is no problem, why not return the same error as SOUP?

```
ResourceError(API::Error::webKitDownloadErrorDomain(), API::Error::Download::Destination, response.url(), localizedDescription);
```

Reply:
> CURLE_WRITE_ERROR does not represent the correct error.

Can you explain why you think so?

From the [curl documentation](https://curl.se/libcurl/c/libcurl-errors.html):
```
CURLE_WRITE_ERROR (23)

An error occurred when writing received data to a local file, or an error was returned to libcurl from a write callback.
```
Also it is the same error code that is [already used](https://github.com/WebKit/WebKit/blob/3f549dfda7946f5b554aa8c45da58e0fb97871aa/Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.cpp#L203) to report similar download errors in webkit.

>  If there is no problem, why not return the same error as SOUP?

There is a problem, `downloadDestinationError` as well as `API::Error::webKitDownloadErrorDomain()` are only available in SOUP (guarded by `USE(SOUP)`).

Reply:
> Can you explain why you think so?

Because it's not an error caused by curl. However, I understand that CURLE_WRITE_ERROR is already used in some places.

NetworkDataTaskBlob seems to return canceledError() in similar cases, or an error of its own. It's not a strong opinion, but why not consider these methods as well?

https://github.com/WebKit/WebKit/blob/29adb2b287fb17f11ff8f085c40b543b90166907/Source/WebKit/NetworkProcess/NetworkDataTaskBlob.cpp#L439-L443

https://github.com/WebKit/WebKit/blob/29adb2b287fb17f11ff8f085c40b543b90166907/Source/WebKit/NetworkProcess/NetworkDataTaskBlob.cpp#L528

https://github.com/WebKit/WebKit/blob/29adb2b287fb17f11ff8f085c40b543b90166907/Source/WebKit/NetworkProcess/NetworkDataTaskBlob.h#L75-L82